feat(blog): Intro to DDD and Core Concepts (#24)
SAP/curated-resources-for-domain-driven-design: blog/0001-ddd-introduction.md

# An Introduction to Domain-Driven Design

![Dall-E Generated Artwork](./images/13dd1691-ac46-465d-b164-96d8c77b38be.png)

> You’ve undoubtedly heard the phrase “Domain-Driven Design” (DDD) thrown around a lot recently as a software development architect. You might be pondering what it is and why it is gaining popularity in business. So, in this blog piece, I’ll introduce you to domain-driven design, describe its advantages, and show you how it can help teams create better software. First, let us describe the term “domain.” A domain in software development alludes to a particular area or subject matter that a system is designed to address. If you were creating an e-commerce system, the name would be "e-commerce." The domain is significant because it sets the system’s needs as well as the business principles that govern it.

Reviewer: Proposal to rephrase:
"So, in this blog piece," => "In this blog post"

> You’ve undoubtedly heard the phrase “Domain-Driven Design” (DDD) thrown around a lot recently as a software development architect. You might be pondering what it is and why it is gaining popularity in business. So, in this blog piece, I’ll introduce you to domain-driven design, describe its advantages, and show you how it can help teams create better software. First, let us describe the term “domain.” A domain in software development alludes to a particular area or subject matter that a system is designed to address. If you were creating an e-commerce system, the name would be "e-commerce." The domain is significant because it sets the system’s needs as well as the business principles that govern it.
> 

Reviewer 2: ```suggestion
You’ve undoubtedly heard the phrase “Domain-Driven Design” (DDD) thrown around a lot recently as a software development architect. You might be pondering what it is and why it is gaining popularity in business. So, in this blog piece, I’ll introduce you to Domain-Driven Design, describe its advantages, and show you how it can help teams create better software. First, let us describe the term "domain". A domain in software development describes a particular area or subject matter that a system is designed to address. If you were creating an e-commerce system, the domain would be "e-commerce." The domain is significant because it sets the system’s needs as well as the business principles that govern it.

```

Reviewer 2: I find "alludes to" too flowery and would rather use "describes" or "represents".

DDD puts the topic at the heart of the creative process. This indicates that figuring out the topic and modeling it in the software is the primary goal. Instead of forcing the business to conform to the constraints of the software, the objective is to develop a software system that represents the business area.

> One of the major advantages of DDD is that it aids in improving software development by fostering a common grasp of the domain. The development team can better grasp the business requirements and make sure the software satisfies those needs by involving subject specialists in the process. Better software that is simpler to manage and expand over time may result from this.

Reviewer 2: ```suggestion
One of the major advantages of DDD is that it aids in improving software development by fostering a common grasp of the domain. The development team can better grasp the business requirements and make sure the software satisfies those needs by involving subject specialists in the process. This should result in better software that is simpler to maintain and expand over time.
```

Reviewer 2: "may result" sounds to me like it's an unintentional side-effect.

> One of the major advantages of DDD is that it aids in improving software development by fostering a common grasp of the domain. The development team can better grasp the business requirements and make sure the software satisfies those needs by involving subject specialists in the process. Better software that is simpler to manage and expand over time may result from this.

Reviewer: Proposal to rephrase:
"team can better grasp" => team gets a better understanding"

"Better software that is simpler to manage and expand over time **will** result from this"

> DDD also aids teams in disassembling complicated systems into smaller, easier-to-manage components. This is accomplished through the idea of "bounded contexts," which are autonomous elements of the system with a clearly specified limit and well-defined interfaces for interacting with other contexts. Teams can escape the “big ball of mud” issue, where the system becomes too complicated to comprehend and manage, by using bound contexts.

Reviewer 2: ```suggestion
DDD also aids teams in decomposing complicated systems into smaller, easier-to-manage components. This is accomplished through the idea of "bounded contexts," which are autonomous elements of the system with a clearly specified limit and well-defined interfaces for interacting with other contexts. Teams can escape the “big ball of mud” issue, where the system becomes too complicated to comprehend and manage, by using bound contexts.
```

Reviewer 2: Nitpick: I find "decomposing" to be the more fitting word. 

A more important point (that is not reflected in my suggestion) is that the way it's currently phrased could be understood as "there already exists a system that is split into smaller parts after the fact", which is only part of what I assume you had in mind here. Maybe something like: "DDD also aids teams in designing a system as composition of smaller subsystems that are aligned with the business domain." Not really 100% a fan of my suggestion as well, but I can't think of anything better right now.

> You should be acquainted with a few fundamental ideas before beginning to use DDD. The idea of "aggregates," which are collections of objects that are handled as a singular entity, is one of them. In order to impose consistency within the domain and guarantee that business standards are observed, aggregates are used.

Reviewer 2: > You should be acquainted with a few fundamental ideas before beginning to use DDD.

Minor: I dislike this phrasing somewhat, since it sounds a little like gatekeeping, at least to me (opinions will differ). Also I'm not sure whether it's even true or helpful to give the advice "you need to understand these x things before you can use DDD". I think especially Event Storming is something you can use and benefit from without looking at the theory before. I would even argue that the theory is easier to grasp after doing event storming. 

Suggestion: "There are a few fundamental ideas that are essential to understanding DDD." 

I'm aware that this is somewhat nitpicky 🙃 

The more important point is that I don't aggregates is one of them. Or at least not in the sense of the code pattern aggregate. Understanding business invariants is important, but a) I don't think you need to teach aggregates to make that point, b) I have not yet had seen a scenario in which a business invariant invalidated an existing context boundary (such scenarios may exist, but I haven't seen one yet) and therefore find it hard to stand behind the point that aggregates are important for strategic design, and c) I understood Eric Evan's statement that he wished he would have changed the order of the blue book to include aggregates as one of the things he wished he would've put into the second part.

> You should be acquainted with a few fundamental ideas before beginning to use DDD. The idea of "aggregates," which are collections of objects that are handled as a singular entity, is one of them. In order to impose consistency within the domain and guarantee that business standards are observed, aggregates are used.

Author: Removed aggregates, et al, jist mentioning bounded context

The idea of an “ubiquitous language” that is used by both coders and subject specialists, is another crucial one. The software will more closely represent the business domain as a result of everyone having a shared knowledge of the domain.

> You should also be acquainted with a number of other crucial words and ideas, including “domain events," “domain services," "repositories," and more. Each of these ideas contributes to domain-driven design in a significant way and aids in making sure the software is well-made and simple to manage.

Reviewer: See comment before. I would leave this part out also the references toa aggregates and detailed concepts in the next section.

> You should also be acquainted with a number of other crucial words and ideas, including “domain events," “domain services," "repositories," and more. Each of these ideas contributes to domain-driven design in a significant way and aids in making sure the software is well-made and simple to manage.

Reviewer 2: ```suggestion
You should also be acquainted with a number of other crucial words and ideas, including “domain events," “domain services," "repositories," and more. Each of these ideas contributes to domain-driven design in a significant way and aids in making sure the software is well-made and simple to maintain.
```

Reviewer 2: "well-made" can mean a lot of things. I believe I get your point, but if possible, I'd try to come up with a more precise phrasing.

Also regarding repositories see previous point about strategic vs tactic deisgn.

> Finally, DDD is a software development design discipline that puts the domain at the center of the design process. By fostering a common knowledge of the subject and disassembling complicated systems into simpler, more manageable components, it aids teams in producing better software. Understanding fundamental ideas like aggregates and bounded contexts, as well as ubiquitous language and other significant words and concepts, is crucial to getting started with DDD. So why not give it a shot and see if it can help in the development of improved software by you and your team?

Reviewer 2: ```suggestion
Finally, DDD is a software design discipline that puts the domain at the center of the design process. By fostering a common understanding of the subject and decomposing complicated systems into simpler, more manageable components, it aids teams in producing better software. Understanding fundamental ideas like aggregates and bounded contexts, as well as ubiquitous language and other significant words and concepts, is crucial to getting started with DDD. So why not give it a shot and see if it can help in the development of improved software by you and your team?
```
